airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialReadUtil.java

```diff
@@ -0,0 +1,233 @@
+        emittedAt,
+        false);
+
+    // This starts processing the binglogs as soon as initial sync is complete, this is a bit different from the current cdc syncs.
+    // We finish the current CDC once the initial snapshot is complete and the next sync starts processing the binlogs
+    return Collections.singletonList(
+        AutoCloseableIterators.concatWithEagerClose(
+            Stream
+                .of(initialLoadIterator, Collections.singletonList(AutoCloseableIterators.lazyIterator(incrementalIteratorSupplier, null)))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList()),
+            AirbyteTraceMessageUtility::emitStreamStatusTrace));
+  }
+
+  /**
+   * Determines the streams to sync for initial primary key load. These include streams that are (i) currently in primary key load (ii) newly added
+   * incremental streams.
+   */
+  public static InitialLoadStreams streamsForInitialPrimaryKeyLoad(final CdcStateManager stateManager, final ConfiguredAirbyteCatalog fullCatalog) {
+    final AirbyteStateMessage airbyteStateMessage = stateManager.getRawStateMessage();
+    final Set<AirbyteStreamNameNamespacePair> streamsStillinPkSync = new HashSet<>();
+
+    // Build a map of stream <-> initial load status for streams that currently have an initial primary key load in progress.
+    final Map<AirbyteStreamNameNamespacePair, PrimaryKeyLoadStatus> pairToInitialLoadStatus = new HashMap<>();
+    if (airbyteStateMessage != null && airbyteStateMessage.getGlobal() != null && airbyteStateMessage.getGlobal().getStreamStates() != null) {
+      airbyteStateMessage.getGlobal().getStreamStates().forEach(stateMessage -> {
+        final JsonNode streamState = stateMessage.getStreamState();
+        final StreamDescriptor streamDescriptor = stateMessage.getStreamDescriptor();
+        if (streamState == null || streamDescriptor == null) {
+          return;
+        }
+
+        if (streamState.has(STATE_TYPE_KEY)) {
+          if (streamState.get(STATE_TYPE_KEY).asText().equalsIgnoreCase(PRIMARY_KEY_STATE_TYPE)) {
+            final PrimaryKeyLoadStatus primaryKeyLoadStatus = Jsons.object(streamState, PrimaryKeyLoadStatus.class);
+            final AirbyteStreamNameNamespacePair pair = new AirbyteStreamNameNamespacePair(streamDescriptor.getName(),
+                streamDescriptor.getNamespace());
+            pairToInitialLoadStatus.put(pair, primaryKeyLoadStatus);
+            streamsStillinPkSync.add(pair);
+          }
+        }
+      });
+    }
+
+    final List<ConfiguredAirbyteStream> streamsForPkSync = new ArrayList<>();
+    fullCatalog.getStreams().stream()
+        .filter(stream -> streamsStillinPkSync.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream())))
+        .map(Jsons::clone)
+        .forEach(streamsForPkSync::add);
+    final List<ConfiguredAirbyteStream> newlyAddedStreams = identifyStreamsToSnapshot(fullCatalog, stateManager.getInitialStreamsSynced());
```

Comment:
Do we care if primary key field type is not suitable for PK load? are there types we want to block?

Reply:
What would be an example? I think in general some types are not as efficient than others - but compared to non-pk fields they are still fast

Reply:
ok. so at the very worst case it might not be efficient.
There is not column type that will break completely (looking at you Polygon geometry type…)

Reply:
I'd be veryyy interested if a user set up a primary index as a Polygon geometry shape